### DISCUSSION TITLE
ch1-basic/ch1-07-error-and-panic

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#614] [General]
# ch1-basic/ch1-07-error-and-panic

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch1-basic/ch1-07-error-and-panic.html

**Comment by erixyuan** [comment]
工程中大量的if err!= handler(); err != nil { xxxx} 这个就感觉很难看了。
虽然golang 的愿意panic 不要随便用，但是做web 开发的时候，在某些逻辑中，我更希望像java 一样throw exception 一样，往外抛出异常。
只要在panic 的时候判断是否是自定义的error，如果是的话，就返回统一异常，如果不是，就继续往外抛出

**Reply by joy717** [reply]
> 工程中大量的if err!= handler(); err != nil { xxxx} 这个就感觉很难看了。
虽然golang 的愿意panic 不要随便用，但是做web 开发的时候，在某些逻辑中，我更希望像java 一样throw exception 一样，往外抛出异常。
只要在panic 的时候判断是否是自定义的error，如果是的话，就返回统一异常，如果不是，就继续往外抛出


所以Java是Java，Go是Go。两者的设计思想是不同的。golang设计的思想就是，有错误的时候应该及时处理。
说到底，还是“错误”跟“异常”是两个概念。 既然是“错误”就应该及时纠正处理，“异常”是程序没有按预期的执行。
Java的这种异常（Exception）方式，会导致最外层必须得捕获所有的Exception，否则必然会出问题。
这两种硬要说哪个更好一些，我是选择Go的处理方式。因为个人也是认为，有错误，应该及时处理，而不是不管不顾的往外面抛。（Java10年开发，Golang5年开发）

**Reply by lucas007007** [reply]
java 15年  go 4年
不谈语言之争。

java的throw有被滥用的可能。如果把throw用在流程控制上，会导致正常的流程也会log出一堆exception信息。
相对而言，if error 满天飞的代码风格，实在是不喜欢，写起来累人。
写go的时候，我觉得我就是跟机器一样一直在敲err，err，err。

**Reply by lucas007007** [reply]
如果go能达到java一样的使用率，码农键盘的ifer四个key得好好加固下。

**Reply by joy717** [reply]
> java 15年 go 4年 不谈语言之争。
> 
> java的throw有被滥用的可能。如果把throw用在流程控制上，会导致正常的流程也会log出一堆exception信息。 相对而言，if error 满天飞的代码风格，实在是不喜欢，写起来累人。 写go的时候，我觉得我就是跟机器一样一直在敲err，err，err。

一开始，个人也不喜欢golang的if err，到处都是.
但后面习惯了也还好。
换个角度来说，这个本来就应该及时处理的。
异常/错误 处理这一块，golang跟java有不同的处理方式，哪个好，见仁见智。我个人没有办法给出一个有说服力的理由。
只能说，我接受了golang的处理方式，也倾向于用这种方式处理。 ：）

**Comment by NingQue0920** [comment]
*上面的代码虽然能够工作，但是隐藏一个 bug。如果第一个 os.Open 调用成功，但是第二个 os.Create 调用失败，那么会在没有释放 src 文件资源的情况下返回。*

函数return之后，函数内部创建的变量不应该也随之释放了吗？为什么会出现bug呢？

**Reply by NingQue0920** [reply]
自己研究了一下，os.Open返回的是一个指针类型变量，指向内存中的File结构体，所以即使src作用域结束，内存中的File结构体并不会被释放。

**Reply by PaiGack** [reply]
正确的做法应该使用 defer 来做清理工作
```go
func CopyFile(dstName, srcName string) (written int64, err error) {
    src, err := os.Open(srcName)
    if err != nil {
        return
    }
    defer  dst.Close()

    dst, err := os.Create(dstName)
    if err != nil {
        return
    }
    defer src.Close()

    written, err = io.Copy(dst, src)
    return
}
```

**Reply by sfx2zh1** [reply]
Google了下 
[stackoverflow](https://stackoverflow.com/questions/58351084/does-go-automatically-close-resources-if-not-explicitly-closed) 以及 [reddit](https://www.reddit.com/r/golang/comments/8lz9fr/why_i_should_defer_close/) 的答案是
go 的 未使用的 File 会被 go GC 回收 
但是回收时 finalizer 不会关闭操作系统上挂的 file descriptor 
而这是个有限资源 leak 多了会影响系统稳定性

**Comment by ZzzJing** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。

**Comment by ZzzJing** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。
